webhooks/unifiedstoragequota/validation/unifiedstoragequota_validator.go

```diff
@@ -0,0 +1,332 @@
+			return CapacityResponse{Response: webhook.Errored(http.StatusBadRequest, errors.New("vm.Spec.Image is required"))}
+		}
+		vmiName := vm.Spec.Image.Name
+		var imageStatus vmopv1.VirtualMachineImageStatus
+
+		switch vm.Spec.Image.Kind {
+		case "VirtualMachineImage":
+			vmi := &vmopv1.VirtualMachineImage{}
+			if err := h.Client.Get(ctx, client.ObjectKey{Namespace: vm.Namespace, Name: vmiName}, vmi); err != nil {
+				if apierrors.IsNotFound(err) {
+					return CapacityResponse{Response: webhook.Errored(http.StatusNotFound, err)}
+				}
+				return CapacityResponse{Response: webhook.Errored(http.StatusInternalServerError, err)}
+			}
+			imageStatus = vmi.Status
+		case "ClusterVirtualMachineImage":
+			cvmi := &vmopv1.ClusterVirtualMachineImage{}
+			if err := h.Client.Get(ctx, client.ObjectKey{Name: vmiName}, cvmi); err != nil {
+				if apierrors.IsNotFound(err) {
+					return CapacityResponse{Response: webhook.Errored(http.StatusNotFound, err)}
+				}
+				return CapacityResponse{Response: webhook.Errored(http.StatusInternalServerError, err)}
+			}
+			imageStatus = cvmi.Status
+		}
+
+		if len(imageStatus.Disks) < 1 || imageStatus.Disks[0].Capacity == nil {
+			return CapacityResponse{Response: webhook.Errored(http.StatusNotFound, errors.New("boot disk not found in image status"))}
+		}
+		capacity = imageStatus.Disks[0].Capacity
+	} else {
+		capacity = resource.NewQuantity(0, resource.BinarySI)
+
+		if vm.Status.Storage.Unshared != nil {
+			capacity.Add(*vm.Status.Storage.Unshared)
+		}
+		// Subtract the PVC usage/limit from the total used value as PVCs
+		// are counted separately.
+		for _, volume := range vm.Status.Volumes {
+			if volume.Type == vmopv1.VirtualMachineStorageDiskTypeManaged {
+				if volume.Used != nil {
+					capacity.Sub(*volume.Used)
+				}
+			}
+		}
+	}
+
+	return CapacityResponse{
+		RequestedCapacity: RequestedCapacity{
+			Capacity:         *capacity,
+			StorageClassName: scName,
+			StoragePolicyID:  sc.Parameters[scParamStoragePolicyID],
+		},
+		Response: webhook.Allowed(""),
+	}
+}
+
+func (h *RequestedCapacityHandler) HandleUpdate(ctx *pkgctx.WebhookRequestContext) CapacityResponse {
+	if !ctx.Obj.GetDeletionTimestamp().IsZero() {
+		return CapacityResponse{Response: admission.Allowed(builder.AdmitMesgUpdateOnDeleting)}
+	}
+	vm := &vmopv1.VirtualMachine{}
+	if err := h.Converter.FromUnstructured(ctx.Obj.UnstructuredContent(), vm); err != nil {
+		return CapacityResponse{Response: webhook.Errored(http.StatusBadRequest, err)}
+	}
+
+	oldVM := &vmopv1.VirtualMachine{}
+	if err := h.Converter.FromUnstructured(ctx.OldObj.UnstructuredContent(), oldVM); err != nil {
+		return CapacityResponse{Response: webhook.Errored(http.StatusBadRequest, err)}
+	}
+
+	if vm.Spec.Advanced != nil && vm.Spec.Advanced.BootDiskCapacity != nil &&
+		oldVM.Spec.Advanced != nil && oldVM.Spec.Advanced.BootDiskCapacity != nil &&
+		vm.Spec.Advanced.BootDiskCapacity.Cmp(*oldVM.Spec.Advanced.BootDiskCapacity) != 0 {
```

Comment:
Should we also address the case when the capacity is added or removed in update requests (i.e. `vm.Spec.Advanced.BootDiskCapacity == nil` or `oldVM.Spec.Advanced.BootDiskCapacity == nil`)?